### DISCUSSION TITLE
Optional substring in steps

### DISCUSSION THREAD

**Discussion by lu-maca** [behave/behave#1099] [Q&A] [answered]
Hi,

is it possible to create a step definition that can contain an optional field and such that returns a boolean, true if the field is present and false if not?

For example:
``` python
@step("I {optional_field:bool} post something")
def step_impl(context, optional_field):
    ...
```
``` gherkin
When I post something
```
would result in `optional_field=False`, but
``` gherkin
When I successfully post something
```
would result in `optional_field=True`.

POSSIBLE SOLUTION:
The only way I've found to mimic this behaviour is the following:

``` python
@step("I {success} post something")
@step("I post something")
def step_impl(context, success=None):
    optional_field = False
    if success == "successfully":
        optional_field = True
```
but this is not really robust, because if I write the step:
``` gherkin
When I try to post something
```
it will match the step above.
However I would like to have a step that accepts only "successfully" as optional field.

**Comment by jenisys** [comment, marked as answer]
You can use optional types by using special registered types (simplified by using the `parse_type` module).

NOTES:

* Your example: If you do not specify a type in a parse-expression, the corresponding regular-expression is normally greedy and will match as much as possible.
* The simplest solution is normally the cardinality-fields expression, like: `{number:Number?}` for an optional number using the `cfparse` matcher. It automatically implicitly generates patterns for other cardinalities by using the registered types/patterns with cardinality=1.

SEE:

* https://behave.github.io/behave.example/datatype/index.html
* https://behave.github.io/behave.example/datatype/cardinality_zero_or_one.html

**Reply by lu-maca** [reply]
Thank you!
